middleware/http/sentinel/middleware.go

```diff
@@ -0,0 +1,165 @@
+// ------------------------------------------------------------
+// Copyright (c) Microsoft Corporation and Dapr Contributors.
+// Licensed under the MIT License.
+// ------------------------------------------------------------
+
+package sentinel
+
+import (
+	"encoding/json"
+	"fmt"
+
+	sentinel "github.com/alibaba/sentinel-golang/api"
+	"github.com/alibaba/sentinel-golang/core/base"
+	"github.com/alibaba/sentinel-golang/core/config"
+	"github.com/dapr/components-contrib/middleware"
+	"github.com/dapr/kit/logger"
+	"github.com/pkg/errors"
+	"github.com/valyala/fasthttp"
+)
+
+type middlewareMetadata struct {
+	AppName string `json:"appName"`
+	// LogConfig
+	LogDir string `json:"logDir"`
+	// Rules
+	FlowRules           string `yaml:"flowRules"`
+	CircuitBreakerRules string `yaml:"circuitBreakerRules"`
+	HotSpotParamRules   string `yaml:"hotSpotParamRules"`
+	IsolationRules      string `yaml:"isolationRules"`
+	SystemRules         string `yaml:"systemRules"`
+}
+
+// NewMiddleware returns a new sentinel middleware
+func NewMiddleware(logger logger.Logger) *Middleware {
+	return &Middleware{logger: logger}
+}
+
+// Middleware is an sentinel middleware
+type Middleware struct {
+	logger logger.Logger
+}
+
+// GetHandler returns the HTTP handler provided by sentinel middleware
+func (m *Middleware) GetHandler(metadata middleware.Metadata) (func(h fasthttp.RequestHandler) fasthttp.RequestHandler, error) {
+	var (
+		meta *middlewareMetadata
+		err  error
+	)
+
+	meta, err = getNativeMetadata(metadata)
+	if err != nil {
+		return nil, errors.Wrap(err, "error to parse sentinel metadata")
+	}
+
+	conf := m.newSentinelConfig(meta)
+	err = sentinel.InitWithConfig(conf)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error to init sentinel with config: %s", conf)
+	}
+
+	err = m.loadSentinelRules(meta)
+	if err != nil {
+		return nil, err
+	}
+
+	return func(h fasthttp.RequestHandler) fasthttp.RequestHandler {
+		return func(ctx *fasthttp.RequestCtx) {
+			resourceName := string(ctx.Method()) + ":" + string(ctx.Path())
+			entry, err := sentinel.Entry(
+				resourceName,
+				sentinel.WithResourceType(base.ResTypeWeb),
+				sentinel.WithTrafficType(base.Inbound),
+			)
+
+			if err != nil {
+				ctx.Error(fasthttp.StatusMessage(fasthttp.StatusTooManyRequests), fasthttp.StatusTooManyRequests)
+
+				return
+			}
+
+			defer entry.Exit()
+			h(ctx)
+		}
+	}, nil
+}
+
+func (m *Middleware) loadSentinelRules(meta *middlewareMetadata) error {
```

Comment:
Is there a good way to test the rules? Both in terms of if they load correctly/incorrectly and if the rules are being applied correctly?

Reply:
added a new test for `loadSentinelRules`